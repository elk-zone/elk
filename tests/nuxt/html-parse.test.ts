import { describe, expect, it } from 'vitest'
import { format } from 'prettier'
import { render as renderTree } from 'ultrahtml'
import type { ContentParseOptions } from '~/composables/content-parse'

describe('html-parse', () => {
  it('empty', async () => {
    const { formatted, serializedText } = await render('')
    expect(formatted).toMatchSnapshot('html')
    expect(serializedText).toMatchSnapshot('text')
  })

  it('link + mention', async () => {
    // https://fosstodon.org/@ayo/109383002937620723
    const { formatted, serializedText } = await render('<p>Happy 🤗 we’re now using <span class="h-card"><a href="https://webtoo.ls/@vitest" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>vitest</span></a></span> (migrated from chai+mocha) <a href="https://github.com/ayoayco/astro-reactive-library/pull/203" rel="nofollow noopener noreferrer" target="_blank"><span class="invisible">https://</span><span class="ellipsis">github.com/ayoayco/astro-react</span><span class="invisible">ive-library/pull/203</span></a></p>')
    expect(formatted).toMatchSnapshot('html')
    expect(serializedText).toMatchSnapshot('text')
  })

  it('custom emoji', async () => {
    const { formatted, serializedText } = await render('Daniel Roe :nuxt:', {
      emojis: {
        nuxt: {
          shortcode: 'nuxt',
          url: 'https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png',
          staticUrl: 'https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png',
          visibleInPicker: true,
        },
      },
    })
    expect(formatted).toMatchSnapshot('html')
    expect(serializedText).toMatchSnapshot('text')
  })

  it('emojis', async () => {
    const { formatted, serializedText } = await render('🇫🇷 👨‍👩‍👦 👩‍🚒🧑🏽‍🚀')
    expect(formatted).toMatchSnapshot('html')
    expect(serializedText).toMatchSnapshot('text')
  })

  it('code frame', async () => {
    // https://webtoo.ls/@antfu/109396489827394721
    const { formatted, serializedText } = await render('<p>Testing code block</p><p>```ts<br />import { useMouse, usePreferredDark } from &#39;@vueuse/core&#39;</p><p>// tracks mouse position<br />const { x, y } = useMouse()</p><p>// is the user prefers dark theme<br />const isDark = usePreferredDark()<br />```</p>')
    expect(formatted).toMatchSnapshot('html')
    expect(serializedText).toMatchSnapshot('text')
  })

  it('code frame 2', async () => {
    const { formatted, serializedText } = await render('<p><span class=\"h-card\"><a href=\"https://webtoo.ls/@antfu\" class=\"u-url mention\">@<span>antfu</span></a></span> Testing<br />```ts<br />const a = hello<br />```</p>')
    expect(formatted).toMatchSnapshot('html')
    expect(serializedText).toMatchSnapshot('text')
  })

  it('inline markdown', async () => {
    const { formatted, serializedText } = await render('<p>text `code` **bold** *italic* ~~del~~</p><p>```js<br />code block<br />```</p>')
    expect(formatted).toMatchSnapshot('html')
    expect(serializedText).toMatchSnapshot('text')
  })

  it('html entities', async () => {
    const { formatted, serializedText } = await render('<p>Hello &lt;World /&gt;.</p>')
    expect(formatted).toMatchSnapshot('html')
    expect(serializedText).toMatchSnapshot('text')
  })
})

async function render(input: string, options?: ContentParseOptions) {
  const tree = parseMastodonHTML(input, options)
  const html = await renderTree(tree)
  let formatted = ''
  const serializedText = treeToText(tree).trim()

  try {
    formatted = format(html, {
      parser: 'html',
    })
  }
  catch (e) {
    formatted = html
  }

  return {
    serializedText,
    input,
    tree,
    html,
    formatted,
  }
}
