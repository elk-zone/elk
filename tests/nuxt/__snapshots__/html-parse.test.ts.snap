// Vitest Snapshot v1

exports[`html-parse > code frame > html 1`] = `
"<p>Testing code block</p><p></p><p><pre><code class=\\"language-ts\\">import { useMouse, usePreferredDark } from '@vueuse/core'
// tracks mouse position
const { x, y } = useMouse()
// is the user prefers dark theme
const isDark = usePreferredDark()</code></pre></p>"
`;

exports[`html-parse > code frame > text 1`] = `
"Testing code block


\`\`\`ts
import { useMouse, usePreferredDark } from '@vueuse/core'
// tracks mouse position
const { x, y } = useMouse()
// is the user prefers dark theme
const isDark = usePreferredDark()
\`\`\`"
`;

exports[`html-parse > code frame 2 > html 1`] = `
"<p>
  <span class=\\"h-card\\"
    ><a
      href=\\"https://webtoo.ls/@antfu\\"
      class=\\"u-url mention\\"
      rel=\\"nofollow noopener noreferrer\\"
      target=\\"_blank\\"
      >@<span>antfu</span></a
    ></span
  >
  Testing<br />
  <pre><code class=\\"language-ts\\">const a = hello</code></pre>
</p>
"
`;

exports[`html-parse > code frame 2 > text 1`] = `
"@antfu Testing

\`\`\`ts
const a = hello
\`\`\`"
`;

exports[`html-parse > custom emoji > html 1`] = `
"Daniel Roe
<picture alt=\\":nuxt:\\" class=\\"custom-emoji\\" data-emoji-id=\\"nuxt\\"
  ><source
    srcset=\\"
      https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png
    \\"
    media=\\"(prefers-reduced-motion: reduce)\\" />
  <img
    src=\\"https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png\\"
    alt=\\":nuxt:\\"
/></picture>
"
`;

exports[`html-parse > custom emoji > text 1`] = `"Daniel Roe :nuxt:"`;

exports[`html-parse > emojis > html 1`] = `
"<img
  src=\\"/emojis/twemoji/1f1eb-1f1f7.svg\\"
  alt=\\"🇫🇷\\"
  class=\\"iconify-emoji iconify-emoji--twemoji\\"
/>
<img
  src=\\"/emojis/twemoji/1f468-200d-1f469-200d-1f466.svg\\"
  alt=\\"👨‍👩‍👦\\"
  class=\\"iconify-emoji iconify-emoji--twemoji\\"
/>
<img
  src=\\"/emojis/twemoji/1f469-200d-1f692.svg\\"
  alt=\\"👩‍🚒\\"
  class=\\"iconify-emoji iconify-emoji--twemoji\\"
/><img
  src=\\"/emojis/twemoji/1f9d1-1f3fd-200d-1f680.svg\\"
  alt=\\"🧑🏽‍🚀\\"
  class=\\"iconify-emoji iconify-emoji--twemoji\\"
/>
"
`;

exports[`html-parse > emojis > text 1`] = `"🇫🇷 👨‍👩‍👦 👩‍🚒🧑🏽‍🚀"`;

exports[`html-parse > empty > html 1`] = `""`;

exports[`html-parse > empty > text 1`] = `""`;

exports[`html-parse > html entities > html 1`] = `
"<p>Hello &lt;World /&gt;.</p>
"
`;

exports[`html-parse > html entities > text 1`] = `"Hello <World />."`;

exports[`html-parse > inline markdown > html 1`] = `"<p>text <code>code</code> <b>bold</b> <em>italic</em> <del>del</del></p><p></p><p><pre><code class=\\"language-js\\">code block</code></pre></p>"`;

exports[`html-parse > inline markdown > text 1`] = `
"text \`code\` **bold** *italic* ~~del~~


\`\`\`js
code block
\`\`\`"
`;

exports[`html-parse > link + mention > html 1`] = `
"<p>
  Happy
  <img
    src=\\"/emojis/twemoji/1f917.svg\\"
    alt=\\"🤗\\"
    class=\\"iconify-emoji iconify-emoji--twemoji\\"
  />
  we’re now using
  <span class=\\"h-card\\"
    ><a
      href=\\"https://webtoo.ls/@vitest\\"
      class=\\"u-url mention\\"
      rel=\\"nofollow noopener noreferrer\\"
      target=\\"_blank\\"
      >@<span>vitest</span></a
    ></span
  >
  (migrated from chai+mocha)
  <a
    href=\\"https://github.com/ayoayco/astro-reactive-library/pull/203\\"
    rel=\\"nofollow noopener noreferrer\\"
    target=\\"_blank\\"
    ><span class=\\"invisible\\">https://</span
    ><span class=\\"ellipsis\\">github.com/ayoayco/astro-react</span
    ><span class=\\"invisible\\">ive-library/pull/203</span></a
  >
</p>
"
`;

exports[`html-parse > link + mention > text 1`] = `"Happy 🤗 we’re now using @vitest (migrated from chai+mocha) https://github.com/ayoayco/astro-reactive-library/pull/203"`;
